osm-fudge/bk_tree.py

```diff
@@ -0,0 +1,42 @@
+'''An implementation of bk-tree
+
+More: https://en.wikipedia.org/wiki/BK-tree
+'''
+
+class BKTree:
+    '''A bk-tree implementation and defined user methods
+       for insertion and querying the string
+    '''
+
+    def __init__(self, root, distanceMetrics, options):
+        self.tree = {root: {}}
+        self.distanceMetrics = distanceMetrics
+        self.options = options
+
+    def insert(self, str1, subtree):
+        self.__insert(str1, subtree)
+
+    def __insert(self, str1, subtree):
+        if not subtree or not subtree.keys() or len(subtree.keys()) == 0:
+            return
+        root = list(subtree.keys())[0]
+        dist = self.distanceMetrics(root, str1, self.options)
+
+        if dist in subtree[root]:
+            self.insert(str1, subtree[root][dist])
+        else:
+            subtree[root][dist] = {str1: {}}
+
+    def lookup(self, str1, subtree, tolerance, results):
+        root = list(subtree.keys())[0]
+        dist = self.distanceMetrics(root, str1, self.options)
+
+        if dist <= tolerance:
+            results.append(root)
+
+        minimum = max([0, dist - tolerance])
+        maximum = dist + tolerance
+
+        for d in range(minimum, maximum + 1):
+            if d in subtree[root]:
+                self.lookup(str1, subtree[root][d], tolerance, results)
```

Comment:
At the moment this is a recursive function. It would be great if we could transform it into an iterative function instead. The benefits we will get from it are: we no longer grow the call stack for every tree level and we can `yield` results which can turn this lookup function into a lazy generator. On top of such a lazy generator we could built functionality like stopping early when we have enough nearest neighbors.

The way to turn a recursive funciton into an iterative function is as follows:
- use a stack (use a Python list and emulate a stack's push/pop functionality with a list's append/pop)
- push the starting tree node on the stack
- while the stack is not empty pop the top item off of the stack, do work on it, push child tree nodes onto the stack